Added mqtts documentation (#6)
nitindermohan/oak-new-docs: content/docs/manuals/networking-internals/mqtt-authentication.md

---
title: "MQTT Authentication"
summary: ""
draft: false
weight: 315
toc: true
seo:
  title: "" # custom title (optional)
  description: "" # custom description (recommended)
  canonical: "" # custom canonical URL (optional)
  noindex: false # false (default) or true
---

## Intra-Cluster Communication

![MQTT Architecture Picture](MQTT-Arch.png)

The above image shows how intra-cluster communication takes place. Nodes send application status and node health reports to the cluster-service-manager.
The cluster-service-manager then appropriately propagates this information to other nodes and the root-service-manager. Since the nodes report sensitive information, such
as IP Addresses, we highly recommend securing the MQTT channels.

{{< callout context="note" title="MQTT" icon="outline/info-circle">}} 

> MQTT is a lightweight messaging Protocol that supports publishing/subscribing to named channels. Oakestra uses MQTT due to it's minimal network usage and low processing overhead. More information at [mqtt.org](https://mqtt.org/).

Reviewer: ```suggestion
MQTT is a lightweight messaging protocol that supports publishing/subscribing to named channels. Oakestra uses MQTT due to its minimal network usage and low processing overhead. More information at [mqtt.org](https://mqtt.org/).
```

{{< /callout >}}

> ## MQTTS

Reviewer: ```suggestion
## MQTTS
MQTT supports exchanging certificates to establish a TLS-secured channel. For this, the server (MQTT Broker) and every client require a certificate-key file pair signed against the same certificate Authority (CA). The MQTT broker can be configured to only accept incoming secured connections and to identify devices by their certificate common name (CN) entry.
```

MQTT supports the exchanging of certificates to establish a TLS secured channel. For this the server (MQTT Broker) and every 
client require a certificate-keyfile pair singed against the same certificate Authority (CA).
The MQTT broker can be configured to only accept incoming secured connection, and to identify devices by their certificate common name (CN) entry.

## Enable Mosquitto Authentication

{{< callout context="caution" title="Requirements" icon="outline/alert-triangle">}} 
* You have a running Oakestra deployment
* You have at least one worker node registered
* (Optional) The NetManager is installed and properly configured
{{< /callout >}}

{{< link-card
  title="Getting Started Guide"
  description="Check out the Getting Started guide to set up your first cluster"
  href="../../../getting-started/oak-environment/create-your-first-oakestra-orchestrator"
  target="_blank"
>}}

### Configuring the Cluster Manager

Navigate into the `cluster_orchestrator` directory in the oakestra repository.

1. Configure the MQTT Broker by adding the following lines to the `mosquitto/mosquitto.conf` file:

>     ```bash

Reviewer: For this concrete case I recommend to use `yml` instead of bash.

This is not really bash nor yaml.
But conf has no nice indentation.

bash has a strange coloring to it here

![image](https://github.com/user-attachments/assets/cd9487f2-e3b3-43c7-b4d7-d023bf66eb69)

where as yml still provides the easy copy button and is homogeneous
![image](https://github.com/user-attachments/assets/071d4f1f-1766-42e4-9da4-c746b6214bea)

            ```
    * **Cluster Service Manager (Client):**
        1. Generate a client key:
            ```bash
            openssl genrsa -aes256 -out cluster_net.key 2048
            ```
        2. Generate a certificate signing request:
            ```bash
            openssl req -out cluster_net.csr -key cluster_net.key -new
            ```
            When prompted for the CN, enter `cluster_service_manager`
        3. Send the CSR to the CA:
            ```bash
            openssl x509 -req -in cluster_net.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out cluster_net.crt -days <duration>
            ```
        4. Export the keyfile password as an environment variable:
            ```bash
            export CLUSTER_SERVICE_KEYFILE_PASSWORD=<keyfile password>
            ```

3. Deploy the cluster with the MQTTS override
    ```bash
    sudo -E docker compose -f docker-compose.yml -f override-mosquitto-auth.yml
    ```

### Configuring a Node
<!--- Subject to change when NodeEngine and NetManager are demonized -->

1. Copy the `ca.crt` and `ca.key` files to the worker node.
2. Generate the certificates
   1. Generate a client key:

>        ```bashopenssl genrsa -aes256 -out client.key 2048

Reviewer: ```suggestion
       ```bash
       openssl genrsa -aes256 -out client.key 2048
```

Otherwise :

![image](https://github.com/user-attachments/assets/4d03db59-c6ec-4e56-aa3f-7377f73dd5c8)

       ```
       When prompted for the CN, enter the public IP of the machine
   3. Send the CSR to the CA:
       ```bash
       openssl x509 -req -in client.csr -CA <path to ca file> -CAkey <path to ca key file> -CAcreateserial -out client.crt -days <duration>
       ```
   4. Decrypt the keyfile:
        ```bash
        openssl rsa -in client.key -out unencrypt_client.key
        ```
   5. Tell your OS to trust the certificate authority by placing the ca.crt file in the `/etc/ssl/certs/` directory
3. Run the NodeEngine:
    ```bash
    sudo NodeEngine -n 0 -p 10100 -a <SYSTEM_MANAGER_URL> -c <path to client.crt> -k <path to unencrypt_client.key>
    ```
4. **(Optional)** Configure the NetManager:
    1. Repeat step 2 for the NetManager
    2. Edit the `/etc/netmanager/netcfg.json` file so that the `"MqttCert"` and `"MqttKey"` fields specify the path to the NetManager certificate and key files
    3. Run the NetManager:
        ```bash
        sudo NetManager -p 10100
        ```

{{< callout context="note" title="Did you know?" icon="outline/rocket">}} The [Oakestra automation repository](https://github.com/oakestra/automation)

> contains many useful scripts such as ones for [creating MQTTS certificate files](https://github.com/oakestra/automation/tree/d43f701134fdf71e1206532883006e1937c38ef9/development_cluster_management/generate_mqtts_certificates). {{< /callout >}}

Reviewer: The end is quite abrupt. It would be great to have 1 sentence or paragraph telling the user that e.g. with this/congrats now you have added MQTTS, or provide an example of how to verify that this is indeed working.

---
title: "MQTT Authentication"
summary: ""
draft: false
weight: 315
toc: true
seo:
  title: "" # custom title (optional)
  description: "" # custom description (recommended)
  canonical: "" # custom canonical URL (optional)
  noindex: false # false (default) or true
---

> # MQTT

Reviewer: Nitpick:

I would use smaller headers (for the entire page - i.e. simply add one '#' more) (i.e. instead of '#', go down to '##') in the body of the markdown, because the title from line 2 (MQTT Authentication) has the same size as the other sections in this page - which IMO looks a bit strange.

E.g:
![image](https://github.com/user-attachments/assets/c9c0791d-becd-48e1-a837-da083eb42c5a)

vs.
![image](https://github.com/user-attachments/assets/d19f0d4b-8656-4d13-a311-c23cc006b1a9)

> # MQTT

Reviewer: At least when I read the header "MQTT" I assume to get information about MQTT - maybe even 1-2 sentences + a link that explains what MQTT is and stands for.

Unfortunately, we do not have the main networking bits in here.
Let's image you are a new user/reader and you are not even familiar with MQTT and you stumble upon this page and read the section with the header MQTT - afterwards you still do not know what this is or what information is send between the nodes.

> The following image shows how intra-cluster communication takes place. Nodes regularly report to the cluster-service-manager,

Reviewer: Suggestions:

1) It feels a bit off to start the entire page by mentioning the image that will come at the end of the paragraph. - I would place this sentence at the end of the paragraph just before showing the figure.

2) The second sentence is quite vague. What gets reported, and what information is propagated?

> and the cluster-service-manager appropiatly propegates this information. Since the nodes report sensitive information, such

Reviewer: ```suggestion
and the cluster-service-manager appropriately propagates this information. Since the nodes report sensitive information, such
```

as IP Addresses, we highly recommend securing the MQTT channels.

![MQTT Architecture Picture](mqtt.png)

# MQTTS

> MQTTS supports the exchanging of certificates to establish a TLS secured channel. For this the server (MQTT Broker) and every 

Reviewer: ```suggestion
MQTTS supports exchanging certificates to establish a TLS-secured channel. For this, the server (MQTT Broker) and every client require a certificate-keyfile pair signed by the same certificate Authority (CA).
The MQTT broker can be configured to only accept incoming secured connections and to identify devices by their certificate CN entry.
```

client require a certificate-keyfile pair singed against the same certificate Authority (CA).
The MQTT broker can be configured to only accept incoming secured connection, and to identify devices by their certificate CN entry.

# Enable Moqsuitto Authentication

{{< callout context="note" title="Did you know?" icon="outline/rocket">}} The [Oakestra automation repository](https://github.com/oakestra/automation)

> contains many useful scripts such as ones for [creating MQTTS certificate files](https://github.com/oakestra/automation/tree/d43f701134fdf71e1206532883006e1937c38ef9/development_cluster_management/generate_mqtts_certificates) {{< /callout >}}

Reviewer: ```suggestion
contains many useful scripts such as ones for [creating MQTTS certificate files](https://github.com/oakestra/automation/tree/d43f701134fdf71e1206532883006e1937c38ef9/development_cluster_management/generate_mqtts_certificates). {{< /callout >}}
```

> The following instructions assume you know how to deploy an Oakestra cluster. Please consult the **Getting Started** section of the the documentation first.

Reviewer: Suggestion:

How about following a similar approach as the other pages and using a requirements callout?
+ Making the "Getting Started" a link?
+ Emphasize on having a running Oakestra Setup instead of "assume you know how to deploy"
~
![image](https://github.com/user-attachments/assets/dece21cc-577b-456d-aa78-b60d84498db0)

## Configuring the Cluster Manager

> Navigate the cluster_orchestrator directory in the oakestra repository.

Reviewer: ```suggestion
Navigate into the `cluster_orchestrator` directory in the oakestra repository.
```

> 1. First configure the MQTT Broker by adding the following lines to mosquitto/mosquitto.conf file

Reviewer: ```suggestion
1. Configure the MQTT Broker by adding the following lines to the `mosquitto/mosquitto.conf` file:
```

          `openssl req -new -x509 -days <duration> -extensions v3_ca -keyout ca.key -out ca.crt`
       2. Generate a server key:\
          `openssl genrsa -out server.key 2048`
       3. Generate a certificate signing request including the URL as a SAN:\
          `openssl req -out server.csr -key server.key -new -addext "subjectAltName = IP:${SYSTEM_MANAGER_URL}, DNS:mqtts"`\
           When prompted for the CN, enter `mqtts`
       4. Send the CSR to the CA\
           `openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days <duration> -copy_extensions copyall`
       5. Grant permissions to read the server keyfile:\
            `chmod 0644 server.key`
    * **Cluster Manager (Client):**
        1. Generate a client key:\
            `openssl genrsa -aes256 -out cluster.key 2048`
        2. Generate a certificate signing request:\
            `openssl req -out cluster.csr -key cluster.key -new`\
            When prompted for the CN, enter `cluster_manager`
        3. Send the CSR to the CA:
            `openssl x509 -req -in cluster.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out cluster.crt -days <duration>`

>         4. Export the keyfile password as an environmental variable:\

Reviewer: https://stackoverflow.com/questions/70677153/are-they-called-environment-or-environmental-variables

```suggestion
        4. Export the keyfile password as an environment variable:\
```

            `export CLUSTER_KEYFILE_PASSWORD=<keyfile password>`
    * **Cluster Service Manager (Client):**
        1. Generate a client key:\
          `openssl genrsa -aes256 -out cluster_net.key 2048`
        2. Generate a certificate signing request:\
          `openssl req -out cluster_net.csr -key cluster_net.key -new`\
          When prompted for the CN, enter `cluster_service_manager`
        3. Send the CSR to the CA:
          `openssl x509 -req -in cluster_net.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out cluster_net.crt -days <duration>`

>         4. Export the keyfile password as an environmental variable:\

Reviewer: ```suggestion
        4. Export the keyfile password as an environment variable:\
```

          `export CLUSTER_SERVICE_KEYFILE_PASSWORD=<keyfile password>`

3. Deploy the cluster with the MQTTS override
`sudo -E docker compose -f docker-compose.yml -f override-mosquitto-auth.yml`

## Configuring a Node

> 1. Copy the ca.crt and ca.key file to node machine.

Reviewer: ```suggestion
1. Copy the `ca.crt` and `ca.key` files to the worker node.
```

2. Generate the certificates
   1. Generate a client key:\
       `openssl genrsa -aes256 -out client.key 2048`
   2. Generate a certificate signing request:\
       `openssl req -out client.csr -key client.key -new`\

>        When prompted for the CN, enter the IP of the machine

Reviewer: ```suggestion
       When prompted for the CN, enter the public IP of the machine
```
